### PR TITLE
Implement UDP networking layer

### PR DESCRIPTION
**Pull request by Bidski**
The purpose of this PR is to 
 - Use [dgram](https://nodejs.org/api/dgram.html) to listen to multiple multicast network addresses 
 - Use WebSocketServer and socket.io to communicate received UDP packets to the browser

The proposed protocol amounts to < 300 bytes when serialised, this should easily fit into a single UDP message packet so we should not need to worry about joining fragmented packets.

Aiming to follow the same pattern that we are using in NUsight2, but replacing NUClearNet.js with dgram.

### CODE REVIEW

File: src/client/network/network.ts
Comment (BrendanAnnable) [inline src/client/network/network.ts]: Why reset the transports on a reconnect?

File: src/client/network/network.ts
Comment (BrendanAnnable) [inline src/client/network/network.ts]: Unused?

File: src/server/network/network.ts
Comment (BrendanAnnable) [inline src/server/network/network.ts]: This means each browser client opens a new UDP socket.

IIUC we could/should just share the same socket among all clients?

File: src/server/network/network.ts
Comment (BrendanAnnable) [inline src/server/network/network.ts]: Sockets will handle serialization automatically, so `JSON.{stringify,parse}` shouldn't be necessary

File: src/server/prod.ts
Comment (BrendanAnnable) [inline src/server/prod.ts]: How about:

```suggestion
const blueTeamAddress = args.blue_addr ?? '239.226.152.163';
```

File: src/server/prod.ts
Comment (BrendanAnnable) [inline src/server/prod.ts]: Where do these defaults come from?

File: src/client/network/network.ts
Comment (Bidski) [reply src/client/network/network.ts]: https://socket.io/docs/client-api/#With-websocket-transport-only

File: src/client/network/network.ts
Comment (Bidski) [reply src/client/network/network.ts]: Not sure what to do with it, but felt we probably wanted to respond to it at some point

**Review by TrentHouliston — CHANGES_REQUESTED** [review]
This thing is so overcomplicated for what you need. Since we're not handling all the complexity of NUsight we don't need all the connection logic that's in here. You need a UDP connection that's listening on a port, and forward that to all socket.io clients you currently have.

File: src/server/network/network.ts
Comment (Bidski) [reply src/server/network/network.ts]: How do we do that?

File: src/server/prod.ts
Comment (Bidski) [reply src/server/prod.ts]: I think I stole that one from NUClear, the other is just +1. No idea if there is anything special about them except that they are in a good range of IPs [https://en.wikipedia.org/wiki/Multicast_address#IPv4](https://en.wikipedia.org/wiki/Multicast_address#IPv4)

File: src/client/main.tsx
Comment (TrentHouliston) [inline src/client/main.tsx]: Didn't we set up NUsight to be WebSocket only? We're unlikely to encounter a browser that doesn't support it these days.

File: src/client/main.tsx
Comment (TrentHouliston) [inline src/client/main.tsx:22]: In general, using any is bad, do you know what type the object you get is?

File: src/client/main.tsx
Comment (TrentHouliston) [inline src/client/main.tsx:26]: Should probably remove all the log lines before we eventually merge

File: src/server/dev.ts
Comment (TrentHouliston) [inline src/server/dev.ts]: Likely should make these ports configurable as well as addresses

File: src/server/dev.ts
Comment (TrentHouliston) [inline src/server/dev.ts]: I feel like having multiple servers listening on an address/port each makes more sense. Also, there are no ports going in here which is super suspicious

File: src/server/dev.ts
Comment (TrentHouliston) [inline src/server/dev.ts]: What is the `any` array that goes here? Does it have a type?

File: src/server/dev.ts
Comment (TrentHouliston) [inline src/server/dev.ts]: If you have a reason string here, why not use it?

File: src/server/network/network.ts
Comment (TrentHouliston) [inline src/server/network/network.ts]: This should be configurable

File: src/server/network/network.ts
Comment (TrentHouliston) [inline src/server/network/network.ts]: I'd take the effort to make it support ipv6 if it's not too hard now rather than do it later

File: src/server/network/network.ts
Comment (TrentHouliston) [inline src/server/network/network.ts]: Does dgram support doing DNS lookups? or do you need to do a DNS lookup here

File: src/server/network/network.ts
Comment (TrentHouliston) [inline src/server/network/network.ts]: I'm pretty sure we would want this to support both broadcast and multicast. So add some checks and don't join the multicast group if you're doing a broadcast address (checking if multicast is pretty easy)

File: src/server/network/network.ts
Comment (TrentHouliston) [inline src/server/network/network.ts]: This should be a this.client.values().forEach(emit_cb=>{})` instead, for some reason javascript people don't like these kinds of loops (ask @BrendanAnnable why I don't know)

File: src/server/network/network.ts
Comment (TrentHouliston) [inline src/server/network/network.ts]: This is gross if you are going to basically write a function that's a constructor just make a new Class.
Or more likely do what I said earlier and make one of these objects be for only a single UDPServer

File: src/server/network/network.ts
Comment (TrentHouliston) [inline src/server/network/network.ts]: Make this be for a single ip/port combination

File: src/server/network/network.ts
Comment (TrentHouliston) [inline src/server/network/network.ts]: I'd make this configurable too. Could be useful if people want to setup this server instead to have a multicast address/port per team like "NUbots" rather than just red and blue

File: src/server/prod.ts
Comment (TrentHouliston) [inline src/server/prod.ts]: I would change the arguments to be more like `'NUbots:239.226.152.162:12345'` or `'Blue Team:239.226.152.162:12345'` and just give a bunch of those instead of hard coded values

File: src/server/prod.ts
Comment (TrentHouliston) [inline src/server/prod.ts]: This could be a list of connections instead which I think would work better

**Review by TrentHouliston — CHANGES_REQUESTED** [review]
Lots of changes could make the architecture of this much cleaner

File: src/client/main.tsx
Comment (Bidski) [reply src/client/main.tsx:26]: Probably not a bad idea

File: src/client/main.tsx
Comment (Bidski) [reply src/client/main.tsx:22]: The socket.io emit function signature is `emit(event: string, ...args: any[])`. I figured we should keep it the same to allow us the most flexibility in what we send. We could probably be more specific on the type once we have decided what it is we want to send, but its going to get cast to an `any` anyway.

File: src/server/network/network.ts
Comment (JosephusPaye) [reply src/server/network/network.ts]: Nah this is fine (IMO). No need to add a function.

`.forEach()` is preferable to a `for` loop where you have to keep track of an index and use it to lookup the current item, but the new `for ... of` doesn't have that problem, and doesn't need an extra function.

Comment (Bidski) [reply src/server/network/network.ts]: so `for ... of` is better in this instance?

File: src/server/dev.ts
Comment (TrentHouliston) [inline src/server/dev.ts]: This should be a .map operation rather than a forEach (since you're making a list from a list)
Also never use var in javascript anymore, use `let` or in this case `const` (since you're not assigning anything else to this variable)

File: src/server/network/network.ts
Comment (TrentHouliston) [inline src/server/network/network.ts]: Do we actually need to know this after we have created it?

File: src/server/network/network.ts
Comment (TrentHouliston) [inline src/server/network/network.ts]: This should just be a type with a constructor rather than an initaliseListener function. Any time you have a function that is`initialiseSomething` it's an indication that your classes are leaking into eachother

**Review by TrentHouliston — CHANGES_REQUESTED** [review]
More review

File: src/server/network/network.ts
Comment (Bidski) [reply src/server/network/network.ts]: It simplifies the logic for determining broadcast vs multicast

File: src/server/network/network.ts
Comment (TrentHouliston) [inline src/server/network/network.ts]: Is there a reason that you made a variable equal to a function rather than a member function? There are some reasons to do that but I think in this case it should just be a normal function
```suggestion
  private onMessage(
    team: TeamData,
    packet: Buffer,
    rinfo: dgram.RemoteInfo,
  ) {
```

File: src/server/network/network.ts
Comment (TrentHouliston) [inline src/server/network/network.ts]:
```suggestion
          let ipv4 = addresses.filter(v => v.family === 4);
          let ipv6 = addresses.filter(v => v.family === 6);
```

File: src/server/network/network.ts
Comment (TrentHouliston) [inline src/server/network/network.ts]:
```suggestion
          if (ipv4.length > 0 || ipv6.length > 0) {
            this.servers.push(
              UDPServer.of(
                { name: team.name, address: ipv4.length > 0 ? ipv4[0] : ipv6[0], port: team.port },
                (team: TeamData, payload: Buffer, rinfo: dgram.RemoteInfo) => {
                  this.onMessage(team, payload, rinfo);
                },
                false,
              ),
            );
```
And don't bother storing ipv4 vs ipv6, just check what the address is later to do the multicast checking

File: src/server/network/network.ts
Comment (TrentHouliston) [inline src/server/network/network.ts]: This isn't a valid way to check, IPv6 addresses might look like FF:00::1, or even FF::1 is a valid multicast address

File: src/server/network/network.ts
Comment (TrentHouliston) [inline src/server/network/network.ts]: This seems like adding state to the IP just to make this if statement easier.
I mean you could literally just do team.address.includes(":")

File: src/server/network/network.ts
Comment (TrentHouliston) [inline src/server/network/network.ts]: Why toUpperCase when you can just check it starts with 'e' instead

**Review by TrentHouliston — CHANGES_REQUESTED** [review]

File: src/server/network/network.ts
Comment (JosephusPaye) [reply src/server/network/network.ts]: Yep

File: src/server/prod.ts
Comment (TrentHouliston) [inline src/server/prod.ts]: You fixed dev, but prod is still the old style

**Review by TrentHouliston — DISMISSED** [review]

File: src/server/network/network.ts
Comment (JosephusPaye) [inline src/server/network/network.ts]: These could be `import ... from`:

```suggestion
import dgram from 'dgram';
```

File: src/server/prod.ts
Comment (JosephusPaye) [inline src/server/prod.ts]: This function could be:

```suggestion
  const [name, address, port] = team.split(':');
  return { name, address, port: Number(port) };
```

Comment (Bidski) [reply src/server/prod.ts]: We are allowing for IPv6 addresses which likely contain `:` so we need to specifically look for the first and last occurrences, or choose a different separator

Comment (JosephusPaye) [reply src/server/prod.ts]: Ah, overlooked that. Perhaps `|` as a separator? Unlikely to be in a team name.

Comment (Bidski) [reply src/server/prod.ts]: `/` is also an option that I thought of

Comment (Bidski) [reply src/server/prod.ts]: `:` was chosen because it started with IPv4 and thats common for `<address>:<port>` notation. It seems that IPv6 also uses `:` to separate the port, but the address is then wrapped in square brackets to disambiguate `[address]:port`

Comment (TrentHouliston) [reply src/server/prod.ts]: The other option is we let the shell separate the tokens and use spaces, every 3 values is a name, IP, port

Comment (Bidski) [reply src/server/prod.ts]: Except its not uncommon for team names to have spaces

Comment (TrentHouliston) [reply src/server/prod.ts]: That's what quotes are for

Comment (Bidski) [reply src/server/prod.ts]: Thats a point, they would need quotes no matter what separator we use

Comment (TrentHouliston) [reply src/server/prod.ts]: But you would no longer have the confusion between the name, IP and port. Because I imagine you could in theory have colons in the team name too

File: src/server/network/network.ts
Comment (JosephusPaye) [inline src/server/network/network.ts]: Prefer the simple syntax for arrays with a single type. The generic version is used when the types are mixed.

```suggestion
  private servers: UDPServer[];
```

File: src/server/network/network.ts
Comment (JosephusPaye) [inline src/server/network/network.ts]:
```suggestion
  constructor(teams: TeamData[]) {
```

File: src/server/network/network.ts
Comment (JosephusPaye) [inline src/server/network/network.ts]:
```suggestion
    this.servers = [];
```

File: src/server/network/network.ts
Comment (JosephusPaye) [inline src/server/network/network.ts]: Optional chaining shouldn't be needed, as `error` is guaranteed to be defined with the `if (error) {` guard.

```suggestion
            throw `UDPServer encountered an error resolving a DNS address: Error: ${error.code}\n${error.stack}`;
```

File: src/server/network/network.ts
Comment (JosephusPaye) [inline src/server/network/network.ts]:
```suggestion
  static of(teams: TeamData[]) {
```

File: src/server/network/network.ts
Comment (JosephusPaye) [inline src/server/network/network.ts]: Just curious, are snake_cased variables used in the rest of this code base?

File: src/server/network/network.ts
Comment (JosephusPaye) [inline src/server/network/network.ts:149]: Do we really need the `of()` pattern here? We use it in NUsight to provide a simple construction interface to classes that need many dependencies to construct. Same with the Network class. Perhaps @BrendanAnnable has some thoughts.

File: .gitignore
Comment (JosephusPaye) [inline .gitignore:24]: What's tags?

File: src/server/server.ts
Comment (JosephusPaye) [inline src/server/server.ts]: `r` is the default when reading, and options can be just the encoding:

```suggestion
    'utf8',
```

File: src/server/server.ts
Comment (JosephusPaye) [inline src/server/server.ts:30]: Perhaps mention that the arguments are one or more of the triplet: `<team name> <address> <port>`.

And probably also document this in the README?

**Review by JosephusPaye — CHANGES_REQUESTED** [review]
Some questions, some suggestions

File: .gitignore
Comment (Bidski) [reply .gitignore:24]: I believe it is a result of ctags, vim kept creating it when I opened some files and it was pissing me off :P

Comment (JosephusPaye) [reply .gitignore:24]: I see. Perhaps a short comment like the others have would be useful.

**Comment by Bidski** [comment]
@JosephusPaye what do we want to do with this?

**Review by JosephusPaye — APPROVED** [review]
I'd say LGTM, unless @BrendanAnnable has something to add.

**Pull request merged**
